src/privatebin.rs

```diff
@@ -36,25 +38,70 @@ pub struct Paste {
     pub ct: String,
     pub meta: Meta,
     pub adata: Data,
+    pub comments: Option<Vec<Comment>>,
 
     #[serde(skip)]
     pub adata_str: String,
 }
 
+impl<'a> Decryptable<'a> for Paste {
+    fn get_ct(&'a self) -> &'a str {
+        &self.ct
+    }
+    fn get_cipher(&'a self) -> &'a Cipher {
+        &self.adata.cipher
+    }
+    fn get_adata_str(&self) -> String {
+        self.adata_str.clone()
+    }
+}
+
+#[derive(Deserialize, Debug, Serialize)]
+pub struct Comment {
+    pub status: Option<i32>,
+    pub id: String,
+    pub pasteid: String,
+    pub parentid: String,
+    pub url: Option<String>,
+    pub v: i32,
+    pub ct: String,
+    pub meta: Meta,
+    pub adata: Cipher,
+
+    #[serde(skip)]
+    pub adata_str: String,
+}
+
+impl<'a> Decryptable<'a> for Comment {
+    fn get_ct(&'a self) -> &'a str {
+        &self.ct
+    }
+    fn get_cipher(&'a self) -> &'a Cipher {
+        &self.adata
+    }
+    fn get_adata_str(&self) -> String {
+        if self.adata_str.is_empty() {
+            serde_json::to_string(&self.adata).unwrap()
+        } else {
+            self.adata_str.clone()
+        }
+    }
+}
+
 #[derive(Deserialize, Debug, Serialize)]
 pub struct Meta {
     pub created: i32,
 }
 
-#[derive(Deserialize, Debug, Serialize)]
+#[derive(Deserialize, Debug)]
```

Comment:
Normal serde serialisation results in a JSON *object* `{ .. }`. But Privatebin expects data (adata) to be a JSON *array* `[ .. ]`. 

PBCLI currently stores `adata_str` as a separate property in paste struct. This was required because direct serialisation of `data` outputs a JSON object string, which is not the original adata string (which is array). 

By using this custom serialisation, we get a json array string identical to the original adata string. 

With this, we can do away with adata_str field too, and the custom `TryFrom<serde_json::Value> for Paste` which sets it.

Reply:
As a demonstration, you can see we haven't added a `TryFrom` for comment. So adata_str value remains `""`.  Line84 is sufficient to reconstruct the correct adata string because of this custom array serializer. 


I was going to propose this way of handling adata, and removal of `adata_str` field in another PR.  But maybe we could resolve this here.

Reply:
Suggested patch, using custom serialize.

```patch
diff --git a/src/api.rs b/src/api.rs
index e41d033..c2b3d52 100644
--- a/src/api.rs
+++ b/src/api.rs
@@ -95,7 +95,7 @@ impl API {
         let status: u32 = value.get("status").unwrap().as_u64().unwrap() as u32;
 
         match status {
-            0 => Ok(value.try_into()?),
+            0 => Ok(serde_json::from_value(value)?),
             1 => Err(PasteError::PasteNotFound),
             s => Err(PasteError::UnknownPasteStatus(s)),
         }
diff --git a/src/privatebin.rs b/src/privatebin.rs
index 90fdb1f..eb00b03 100644
--- a/src/privatebin.rs
+++ b/src/privatebin.rs
@@ -3,9 +3,7 @@ use crate::error::PbResult;
 use serde::ser::{SerializeTuple, Serializer};
 use serde::Deserialize;
 use serde::Serialize;
-use serde_json::Value;
 use serde_with::skip_serializing_none;
-use std::io::ErrorKind;
 use url::Url;
 
 #[derive(Deserialize, Debug, Serialize)]
@@ -39,9 +37,6 @@ pub struct Paste {
     pub meta: Meta,
     pub adata: Data,
     pub comments: Option<Vec<Comment>>,
-
-    #[serde(skip)]
-    pub adata_str: String,
 }
 
 impl<'a> Decryptable<'a> for Paste {
@@ -52,7 +47,7 @@ impl<'a> Decryptable<'a> for Paste {
         &self.adata.cipher
     }
     fn get_adata_str(&self) -> String {
-        self.adata_str.clone()
+        serde_json::to_string(&self.adata).unwrap()
     }
 }
 
@@ -67,9 +62,6 @@ pub struct Comment {
     pub ct: String,
     pub meta: Meta,
     pub adata: Cipher,
-
-    #[serde(skip)]
-    pub adata_str: String,
 }
 
 impl<'a> Decryptable<'a> for Comment {
@@ -80,11 +72,7 @@ impl<'a> Decryptable<'a> for Comment {
         &self.adata
     }
     fn get_adata_str(&self) -> String {
-        if self.adata_str.is_empty() {
-            serde_json::to_string(&self.adata).unwrap()
-        } else {
-            self.adata_str.clone()
-        }
+        serde_json::to_string(&self.adata).unwrap()
     }
 }
 
@@ -229,21 +217,3 @@ impl Serialize for Cipher {
         s.end()
     }
 }
-
-impl TryFrom<serde_json::Value> for Paste {
-    type Error = crate::error::PbError;
-
-    fn try_from(value: Value) -> PbResult<Self> {
-        let adata = value.get("adata").ok_or(std::io::Error::new(
-            ErrorKind::InvalidData,
-            "Cannot get adata in try_from",
-        ))?;
-        let adata_str = serde_json::to_string(adata)?;
-
-        let mut paste = serde_json::from_value::<Paste>(value)?;
-
-        paste.adata_str = adata_str;
-
-        Ok(paste)
-    }
-}

```

Reply:
Hey,

yea - with the removal of adata_str that makes a lot more sense now :), and I like the idea.

Do you want to add that change as a commit in this PR? 

In addition, did you see the lifetime comment I made above?

Reply:
> In addition, did you see the lifetime comment I made above?

I saw it only now, sorry for missing it earlier.  Didn't realise the compiler would figure it out, so I started explicit, and then just carried on.  I'll remove them :)

> Do you want to add that change as a commit in this PR?

Sure, I can it here, unless you want it as separate PR. If so, let me know.

Reply:
> I saw it only now, sorry for missing it earlier. Didn't realise the compiler would figure it out, so I started explicit, and then just carried on. I'll remove them :)

Cool!

> Sure, I can it here, unless you want it as separate PR. If so, let me know.

Yeah, lets add it here, so its coupled together with the added custom serialization.  Apart from that it LGTM. When the Lifetime fix is made and the removal of adata_str commit is in, I would proceed to merge it.

Thanks for the contribution!

Best Regards
Mydayyy